File: Examples/EmojiFun/bindings/kotlin/src/test/kotlin/com/cricut/emojifun/EmojiExplorerTests.kt
package com.cricut.emojifun

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

internal class EmojiExplorerTests {
    @Test
    fun hello() {
        assertEquals(EmojiExplorer.hello, "👋🌎")
    }

    @Test
    fun construction() {
        val explorer = EmojiExplorer.create(128)
        assertEquals(explorer.known.count(), 128)
        assertEquals(explorer.known, hashSetOf(
            "⌚", "⌛", "⏩", "⏪", "⏫", "⏬", "⏰", "⏳", "◽", "◾", "☔", "☕", "♈", "♉", "♊", "♋",
            "♌", "♍", "♎", "♏", "♐", "♑", "♒", "♓", "♿", "⚓", "⚡", "⚪", "⚫", "⚽", "⚾", "⛄",
            "⛅", "⛎", "⛔", "⛪", "⛲", "⛳", "⛵", "⛺", "⛽", "✅", "✊", "✋", "✨", "❌", "❎", "❓",
            "❔", "❕", "❗", "➕", "➖", "➗", "➰", "➿", "⬛", "⬜", "⭐", "⭕", "🀄", "🃏", "🆎", "🆑",
            "🆒", "🆓", "🆔", "🆕", "🆖", "🆗", "🆘", "🆙", "🆚", "🇦", "🇧", "🇨", "🇩", "🇪", "🇫", "🇬",
            "🇭", "🇮", "🇯", "🇰", "🇱", "🇲", "🇳", "🇴", "🇵", "🇶", "🇷", "🇸", "🇹", "🇺", "🇻", "🇼",
            "🇽", "🇾", "🇿", "🈁", "🈚", "🈯", "🈲", "🈳", "🈴", "🈵", "🈶", "🈸", "🈹", "🈺", "🉐", "🉑",
            "🌀", "🌁", "🌂", "🌃", "🌄", "🌅", "🌆", "🌇", "🌈", "🌉", "🌊", "🌋", "🌌", "🌍", "🌎", "🌏",
        ))

        assertTrue(EmojiExplorer.create(0).known.isEmpty())
        assertTrue(EmojiExplorer.create(-1).known.isEmpty())
        assertTrue(EmojiExplorer.create(1212 + 50).known.count() >= 1212)
    }

    @Test
    fun random() {
        val explorer = EmojiExplorer.create(128)
        for (i in 0..<100) {
            assertFalse(explorer.random().isEmpty())
        }

        assertEquals(EmojiExplorer.create(0).random(), EmojiExplorer.hello)
    }

    @Test
    fun randomUnique() {
        val count = 128L
        val explorer = EmojiExplorer.create(count)

        val reported = ArrayList<String>()
        for (i in 0..<explorer.known.count()) {
            val emoji = explorer.randomUnique()
            if (emoji != null) {
                reported.add(emoji)
            }
        }

        assertEquals(reported.count(), count.toInt())
        assertEquals(reported.toHashSet(), explorer.known)
    }

    @Test
    fun enumerateKnown() {
        val count = 128L
        val explorer = EmojiExplorer.create(count)

        val reported = ArrayList<String>()
        explorer.enumerateKnown() { s ->
            reported.add(s)
            true
        }

        assertEquals(reported.count(), count.toInt())
        assertEquals(reported.toHashSet(), explorer.known)

        var earlyExitReportedCount = 0
        explorer.enumerateKnown() { s ->
            earlyExitReportedCount += 1
            earlyExitReportedCount < 50
        }
        assertEquals(earlyExitReportedCount, 50)
    }
}

File: Examples/EmojiFun/Tests/EmojiFunTests/EmojiExplorerTests.swift
import Testing
import EmojiFun

@Test func hello() throws {
    #expect(!EmojiExplorer.hello.isEmpty)
    #expect(EmojiExplorer.hello.unicodeScalars.allSatisfy { $0.properties.isEmoji })
}

@Test func construction() {
    let explorer = EmojiExplorer.init(count: 128)
    #expect(explorer.known.count == 128)
    #expect(explorer.known.allSatisfy { $0.unicodeScalars.contains { $0.properties.isEmoji } })
    #expect(explorer.known == [
        "⌚", "⌛", "⏩", "⏪", "⏫", "⏬", "⏰", "⏳", "◽", "◾", "☔", "☕", "♈", "♉", "♊", "♋",
        "♌", "♍", "♎", "♏", "♐", "♑", "♒", "♓", "♿", "⚓", "⚡", "⚪", "⚫", "⚽", "⚾", "⛄",
        "⛅", "⛎", "⛔", "⛪", "⛲", "⛳", "⛵", "⛺", "⛽", "✅", "✊", "✋", "✨", "❌", "❎", "❓",
        "❔", "❕", "❗", "➕", "➖", "➗", "➰", "➿", "⬛", "⬜", "⭐", "⭕", "🀄", "🃏", "🆎", "🆑",
        "🆒", "🆓", "🆔", "🆕", "🆖", "🆗", "🆘", "🆙", "🆚", "🇦", "🇧", "🇨", "🇩", "🇪", "🇫", "🇬",
        "🇭", "🇮", "🇯", "🇰", "🇱", "🇲", "🇳", "🇴", "🇵", "🇶", "🇷", "🇸", "🇹", "🇺", "🇻", "🇼",
        "🇽", "🇾", "🇿", "🈁", "🈚", "🈯", "🈲", "🈳", "🈴", "🈵", "🈶", "🈸", "🈹", "🈺", "🉐", "🉑",
        "🌀", "🌁", "🌂", "🌃", "🌄", "🌅", "🌆", "🌇", "🌈", "🌉", "🌊", "🌋", "🌌", "🌍", "🌎", "🌏",
    ])

    #expect(EmojiExplorer.init(count: 0).known.isEmpty)
    #expect(EmojiExplorer.init(count: -1).known.isEmpty)
    #expect(EmojiExplorer.init(count: 1212 + 50).known.count >= 1212)
}

@Test func random() throws {
    let explorer = EmojiExplorer.init(count: 128)
    for _ in 0..<100 {
        #expect(explorer.random().unicodeScalars.contains { $0.properties.isEmoji })
    }

    #expect(EmojiExplorer.init(count: 0).random() == EmojiExplorer.hello)
}

@Test func randomUnique() throws {
    let count = 128
    var explorer = EmojiExplorer.init(count: count)

    var reported: [String] = []
    for _ in 0..<explorer.known.count {
        let emoji = explorer.randomUnique()
        reported.append(try #require(emoji))
    }
    
    #expect(reported.count == count)
    #expect(Set(reported) == explorer.known)
}

@Test func enumerateKnown() throws {
    let count = 128
    let explorer = EmojiExplorer.init(count: count)

    var reported: [String] = []
    explorer.enumerateKnown { s in
        reported.append(s)
        return true
    }

    #expect(reported.count == count)
    #expect(Set(reported) == explorer.known)

    var earlyExitReportedCount = 0
    explorer.enumerateKnown { s in
        earlyExitReportedCount += 1
        return earlyExitReportedCount < 50
    }
    #expect(earlyExitReportedCount == 50)
}
